### ISSUE TITLE
<defs>...</defs> must precede <use ... />

### ISSUE THREAD

**Issue by vladdeSV** [SVGKit/SVGKit#540]
### Issue
Defining (`<defs>...</defs>`) after using (`<use xlink:href="#tag" />`) causes error `Found an SVG <use> tag that points to a non-existent element. Missing element: id = tag`

### Expected
`<defs>...</defs>` can be defined anywhere in the SVG.

### Example

The following SVG-file will cause fail, and crash
```xml
<?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" width="100" height="100">
   <use xlink:href="#rekk"/>
   <defs>
      <rect id="rekk" x="3" y="3" width="14" height="14" fill="red"/>
   </defs>
</svg> 
```

Moving the `<defs>...</defs>` above the `<use ... />` will run and render perfectly.

**Comment by adamgit** [comment]
My initial thoughts are: any software that is writing defs out-of-order is
being deliberately unhelpful (I would go so far as to suggest it's a bug:
I'm struggling to think of any good reason for it, and it has obvious bad
effects). Why would you declare something AFTER you've used it :) ? The
spec doesn't encourage this (all their examples are the sensible way
around: declaration first, usage second).

Off the top of my head: SVGKit uses a streaming parser (fast, small memory
footprint) ... allowing for undeclared forward declarations could be
difficult. I'm sure it's possible, but it would require going from a
single-pass parser to a multi-pass parser (not much of a code change, but
harder to maintain and optimize).

**Comment by vladdeSV** [comment]
I totally agree `<defs>` should be defined before. However, currently it does work with Chrome, and it isn't documented anywhere that definitions must be defined beforehand.

Maybe add a disclaimer that all `<defs>` must come before it's usage?

**Comment by adamgit** [comment]
Sounds like a good idea.

Perhaps the most effective way is to modify the error message, and add the
disclaimer directly there. So if anyone encounters this, they immediately
see both the cause and the simplest solution.

...even better if it references this Issue, so anyone hitting it can come
and comment here when they fix it ;).
